broker/server_apis.go

```diff
@@ -0,0 +1,246 @@
+/***************************************************************
+ *
+ * Copyright (C) 2024, Pelican Project, Morgridge Institute for Research
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you
+ * may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************/
+
+package broker
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/pelicanplatform/pelican/param"
+	"github.com/pelicanplatform/pelican/token_scopes"
+	log "github.com/sirupsen/logrus"
+)
+
+type (
+
+	// Structure for an origin's POST to the broker
+	originRequest struct {
+		Origin string `json:"origin"`
+		Prefix string `json:"prefix"`
+	}
+
+	// Structure for an origin retrieval response from the broker
+	originResp struct {
+		Status  string          `json:"status"`
+		Request reversalRequest `json:"request"`
+	}
+
+	// Error response for a request or retrieval
+	brokerErrResp struct {
+		Status string `json:"status"`
+		Msg    string `json:"msg"`
+	}
+
+	// Structure for an origin calling back to the cache
+	callbackRequest struct {
+		RequestId string `json:"request_id"`
+	}
+)
+
+func retrieveRequest(ctx context.Context, ginCtx *gin.Context) {
+	timeoutStr := "5s"
+	if val := ginCtx.Request.Header.Get("X-Pelican-Timeout"); val != "" {
+		timeoutStr = val
+	}
+
+	timeoutVal, err := time.ParseDuration(timeoutStr)
+	if err != nil {
+		ginCtx.String(http.StatusBadRequest, "Failed to parse X-Pelican-Timeout header to a duration (example: 5s)")
+		ginCtx.Abort()
+		return
+	}
+
+	originReq := originRequest{}
+	if err := ginCtx.Bind(&originReq); err != nil {
+		ginCtx.String(http.StatusBadRequest, "Failed to parse the origin's retrieve request")
+		ginCtx.Abort()
+		return
+	}
+
+	token := ginCtx.Request.Header.Get("Authorization")
+	token, hasPrefix := strings.CutPrefix(token, "Bearer ")
+	if !hasPrefix {
+		ginCtx.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"status": "error", "msg": "Bearer authorization required for callback"})
```

Comment:
In many other places, we return error in the form of `{"error":"<error-message>"}` or a success request in `{"msg": "Success"}`

The change in schema might break the consistency (although we are certainly not consistently using the exact same default server response message throughout the codebase)

Reply:
Honestly, I think having a separate status and message field is the better way to go.  This allows us to more easily have a common schema between success and failure responses.

To help keep things portable, I did wrap everything up into a common struct, allowing an easier future refactoring.